File: csv-tools/scripts-node/group.mjs
#!/usr/bin/env node
// Copyright 2026 Yevgeniy Zaremba
// SPDX-License-Identifier: Apache-2.0

import { parseArgs } from "node:util";
import {
  DEFAULT_ENCODING,
  loadCsv,
  parseDelimiter,
  writeJson,
} from "./common.mjs";

function parseFields(raw) {
  return raw.split(",").map((f) => f.trim()).filter(Boolean);
}

function parseAgg(spec) {
  if (!spec.includes(":")) throw new Error(`Invalid --agg '${spec}'. Use field:func.`);
  const idx = spec.lastIndexOf(":");
  const field = spec.slice(0, idx).trim();
  const func = spec.slice(idx + 1).trim().toLowerCase();
  const allowed = new Set(["count", "sum", "min", "max", "mean", "list", "unique"]);
  if (!allowed.has(func)) throw new Error(`Unknown agg: ${func}`);
  return [field, func];
}

function computeAgg(values, func) {
  if (func === "count") return values.length;
  if (func === "list") return values;
  if (func === "unique") {
    const seen = new Set();
    const out = [];
    for (const v of values) {
      if (!seen.has(v)) {
        seen.add(v);
        out.push(v);
      }
    }
    return out;
  }
  const nums = [];
  for (const v of values) {
    const s = (v ?? "").trim();
    if (!s) continue;
    const n = Number.parseFloat(s);
    if (!Number.isNaN(n)) nums.push(n);
  }
  if (!nums.length) return null;
  if (func === "sum") return nums.reduce((a, b) => a + b, 0);
  if (func === "min") return Math.min(...nums);
  if (func === "max") return Math.max(...nums);
  if (func === "mean") return nums.reduce((a, b) => a + b, 0) / nums.length;
  return null;
}

function main() {
  const { values, positionals } = parseArgs({
    options: {
      delimiter: { type: "string", default: "," },
      "no-header": { type: "boolean", default: false },
      "skip-lines": { type: "string" },
      "comment-char": { type: "string" },
      encoding: { type: "string", default: DEFAULT_ENCODING },
      by: { type: "string" },
      agg: { type: "string", multiple: true, default: [] },
      sort: { type: "string", default: "count" },
      top: { type: "string" },
      compact: { type: "boolean", default: false },
    },
    allowPositionals: true,
  });

  if (!values.by) {
    console.error("group.mjs: --by is required");
    process.exit(1);
  }

  const input = positionals[0] ?? "-";
  const { columns, rows } = loadCsv(input, {
    delimiter: parseDelimiter(values.delimiter),
    hasHeader: !values["no-header"],
    commentChar: values["comment-char"] ?? null,
    encoding: values.encoding,
    skipLines: values["skip-lines"] != null ? Number.parseInt(values["skip-lines"], 10) : null,
  });

  let byFields = parseFields(values.by);
  byFields = byFields.filter((f) => columns.includes(f));
  if (!byFields.length) byFields = columns.slice(0, 1);

  const aggSpecs = values.agg.map(parseAgg);

  const groups = new Map();
  for (const row of rows) {
    const key = JSON.stringify(byFields.map((f) => (row[f] ?? "").trim()));
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(row);
  }

  let outRows = [];
  for (const [keyStr, groupRows] of groups) {
    const keyTuple = JSON.parse(keyStr);
    const rowDict = {};
    for (let i = 0; i < byFields.length; i++) rowDict[byFields[i]] = keyTuple[i];
    rowDict.count = groupRows.length;
    for (const [aggField, aggFunc] of aggSpecs) {
      const vals = groupRows.map((r) => (r[aggField] ?? "").trim());
      rowDict[`${aggField}:${aggFunc}`] = computeAgg(vals, aggFunc);
    }
    outRows.push(rowDict);
  }

  if (values.sort === "count") {
    outRows.sort((a, b) => b.count - a.count);
  } else {
    outRows.sort((a, b) => {
      for (const f of byFields) {
        const va = a[f] ?? "";
        const vb = b[f] ?? "";
        if (va !== vb) return va < vb ? -1 : 1;
      }
      return 0;
    });
  }

  if (values.top != null) {
    const n = Math.max(0, Number.parseInt(values.top, 10));
    outRows = outRows.slice(0, n);
  }

  const result = {
    total_records: rows.length,
    total_groups: groups.size,
    groups: outRows,
  };
  writeJson(result, values.compact);
}

main();

File: csv-tools/scripts/group.py
#!/usr/bin/env python3
# Copyright 2026 Yevgeniy Zaremba
# SPDX-License-Identifier: Apache-2.0

"""Group CSV rows by column(s) with optional aggregations."""

from __future__ import annotations

import argparse
from collections import defaultdict

from common import load_csv, parse_delimiter, write_json


def add_dialect_args(parser: argparse.ArgumentParser) -> None:
    parser.add_argument("input", nargs="?", default="-", help="Input CSV file or '-' for stdin.")
    parser.add_argument("--delimiter", default=",", help="Field delimiter (use '\\t' for tab).")
    parser.add_argument("--no-header", action="store_true", help="First row is data.")
    parser.add_argument("--skip-lines", type=int, default=None, help="Skip this many preamble lines; next line is header.")
    parser.add_argument("--comment-char", help="Skip lines starting with this character.")
    parser.add_argument("--encoding", default="utf-8", help="File encoding.")


def parse_fields(raw: str) -> list[str]:
    return [f.strip() for f in raw.split(",") if f.strip()]


def parse_agg(spec: str) -> tuple[str, str]:
    if ":" not in spec:
        raise ValueError(f"Invalid --agg '{spec}'. Use field:func.")
    field, func = spec.rsplit(":", 1)
    func = func.strip().lower()
    if func not in {"count", "sum", "min", "max", "mean", "list", "unique"}:
        raise ValueError(f"Unknown agg: {func}")
    return (field.strip(), func)


def compute_agg(values: list[str], func: str):
    if func == "count":
        return len(values)
    if func == "list":
        return values
    if func == "unique":
        seen: list[str] = []
        seen_set: set[str] = set()
        for v in values:
            if v not in seen_set:
                seen_set.add(v)
                seen.append(v)
        return seen
    nums = []
    for v in values:
        v = (v or "").strip()
        if not v:
            continue
        try:
            nums.append(float(v))
        except ValueError:
            pass
    if not nums:
        return None
    if func == "sum":
        return sum(nums)
    if func == "min":
        return min(nums)
    if func == "max":
        return max(nums)
    if func == "mean":
        return sum(nums) / len(nums)
    return None


def main() -> None:
    parser = argparse.ArgumentParser(description="Group CSV by columns with optional aggregations.")
    add_dialect_args(parser)
    parser.add_argument("--by", required=True, help="Comma-separated columns to group by.")
    parser.add_argument("--agg", action="append", default=[], help="field:func (sum, mean, min, max, list, unique).")
    parser.add_argument("--sort", choices=["count", "key"], default="count", help="Sort groups by count (desc) or key (asc).")
    parser.add_argument("--top", type=int, help="Limit to top N groups.")
    parser.add_argument("--compact", action="store_true", help="Emit compact JSON.")
    args = parser.parse_args()

    columns, rows = load_csv(
        args.input,
        delimiter=parse_delimiter(args.delimiter),
        has_header=not args.no_header,
        comment_char=args.comment_char,
        encoding=args.encoding,
        skip_lines=args.skip_lines,
    )[:2]

    by_fields = parse_fields(args.by)
    by_fields = [f for f in by_fields if f in columns]
    if not by_fields:
        by_fields = columns[:1]

    agg_specs = [parse_agg(s) for s in args.agg]

    groups: dict[tuple, list[dict]] = defaultdict(list)
    for row in rows:
        key = tuple((row.get(f, "") or "").strip() for f in by_fields)
        groups[key].append(row)

    out_rows: list[dict] = []
    for key_tuple, group_rows in groups.items():
        row_dict = dict(zip(by_fields, key_tuple))
        row_dict["count"] = len(group_rows)
        for agg_field, agg_func in agg_specs:
            vals = [(r.get(agg_field, "") or "").strip() for r in group_rows]
            label = f"{agg_field}:{agg_func}"
            row_dict[label] = compute_agg(vals, agg_func)
        out_rows.append(row_dict)

    if args.sort == "count":
        out_rows.sort(key=lambda r: r["count"], reverse=True)
    else:
        out_rows.sort(key=lambda r: tuple(r.get(f, "") for f in by_fields))

    if args.top is not None:
        out_rows = out_rows[: max(0, args.top)]

    result = {"total_records": len(rows), "total_groups": len(groups), "groups": out_rows}
    write_json(result, compact=args.compact)


if __name__ == "__main__":
    main()
